lib/minstall.js

```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
+            delete requestedDependencies[dependency][version];
+          }
 
-      let command = moduleInfo.postinstall;
-      if (command.indexOf('minstall') === 0) {
-        command += ' --isChildProcess';
+          requestedDependencies[dependency][intersection].push(module.fullModulePath);
+          versionFound = true;
+        }
       }
-      return systools.runCommand(`cd ${modulePath}${commandConcatSymbol} ${command}`);
-    })
-    .then(() => {
 
-      return systools.delete(path.join(modulePath, 'node_modules'));
-    });
+      if (!versionFound) {
+        requestedDependencies[dependency][module.dependencies[dependency]] = [
+          module.fullModulePath,
+        ];
+      }
+    }
+  }
+
+  return requestedDependencies;
 }
 
-function cacheConflictingModules(conflictedDependencies, localModules, moduleFolderName) {
+function removeAlreadySatisfiedDependencies(requestedDependencies, localModules, alreadyInstalledDependencies) {
+  const result = Object.assign({}, requestedDependencies);
+  for (const requestedDependencyName in result) {
+    for (const requestedDependencyVersionRange in result[requestedDependencyName]) {
 
-  if (conflictedDependencies.length === 0) {
-    logger.debug('there are no conflicting packages, thus nothing needs to be cached');
-    return Promise.resolve([]);
-  }
+      // check if already installed modules satisfy the dependency
+      for (const installedDependency of alreadyInstalledDependencies) {
+        if (requestedDependencyName !== installedDependency.name) {
+          continue;
+        }
 
-  const localModuleNames = localModules.map((module) => {
-    return module.name;
-  });
+        if (!semver.satisfies(installedDependency.version, requestedDependencyVersionRange)) {
+          continue;
+        }
 
-  // local modules are linked at the very end of the installation, thus all
-  // local modules are "not yet linked"
-  let cachedModuleFolders = conflictedDependencies;
-  if (linkModules) {
-    // don't try to cache away local modules, that will be linked later
-    cachedModuleFolders = cachedModuleFolders.filter((module) => {
-      return localModuleNames.indexOf(module.name) < 0;
-    });
-  }
+        for (const locationOfRequestingModule of result[requestedDependencyName][requestedDependencyVersionRange]) {
+          // It doesn't matter where the dependency with the correct version is! all shadowed dependencies
+          // get fixed with symlinks after the installation!
+          const shortenedRequesterLocation = locationOfRequestingModule.substr(cwd.length);
+          const shortenedInstalledLocation = installedDependency.location.substr(cwd.length);
+          logger.debug(`dependency ${requestedDependencyName}@${requestedDependencyVersionRange} requested by '${shortenedRequesterLocation}' will be satisfied by installed version ${installedDependency.version} in '${path.join(shortenedInstalledLocation)}'`);
```

Comment:
I would plainly just use a log to info and error.

We could consider dropping debug-loglevel entirely.

Reply:
this can log hundrets of lines you most likely need when debugging, but don't care about when using minstall.